### ISSUE TITLE
Please explain in terms of JavaScript API

### ISSUE THREAD

**Issue by domenic** [richtr/html-media-focus#1]
One of the goals of the [Extensible Web](http://extensiblewebmanifesto.org/) is to be able to explain higher-level features like markup in terms of lower-level programmatic JavaScript APIs. This allows developers to gain access to the same abilities that native elements have, without having to e.g. create a dummy `<audio>` or `<video>` just to get access to media controls.

This is being concretely realized in the Google Chrome team's [HTML as Custom Elements](https://github.com/domenic/html-as-custom-elements) project, where we are using custom elements to reimplement the base elements of HTML---like `<video>` or `<audio>`.

So, if we were implementing `<audio>` in JavaScript using custom elements, how would we implement `<audio focusable>`? An API needs to exist to allow this.

**Comment by tabatkins** [comment]
This would also make some interactions more clear - for example, what happens if the play() event is canceled?  Does the element still get media focus?  An API described in terms of JS, that other elements could hook into if desired, would have the answer fall out of the description.

For example, there could be a MediaFocusSet that you can put elements into; the default action of play() is to check if the target is in the set, and if so, do the behavior described here (pause the currently-focused element, play the new one).

To fully explain it, you'd need to solve the Web Audio problem, perhaps in terms of a duck-typed player API that objects have to implement in order to pass the MediaFocusSet typing checks.  Then custom elements could hook into this properly, generating or managing their own audio but still responding to media events automatically.  This might just be as simple as saying "we're gonna fire these media events at you, make sure you can respond to them".  The object could optionally offer its own currentTime/duration/name/icon/etc properties that the platform can observe, for the purposes stated in this proposal.

**Comment by richtr** [comment]
> One of the goals of the Extensible Web is to be able to explain higher-level features like markup in terms of lower-level programmatic JavaScript APIs.

I'm a big fan of the Extensible Web Manifesto \o/ ...

> One of the goals of the Extensible Web is to be able to explain higher-level features like markup in terms of lower-level programmatic JavaScript APIs. This allows developers to gain access to the same abilities that native elements have, without having to e.g. create a dummy &lt;audio&gt; or &lt;video&gt; just to get access to media controls.
> 
> So, if we were implementing <audio> in JavaScript using custom elements, how would we implement <audio focusable>? An API needs to exist to allow this.

Custom elements that want to play media must use one of: &lt;audio&gt;, &lt;video&gt; or `AudioContext` to produce any _audible content_ within a web page. Creating a custom element that doesn't feed in to one of these APIs is not going to produce any audible media output.

We could therefore argue that `HTMLMediaElement` and `AudioContext` provide all the low-level API primitives for media within JavaScript. Is there any way within custom elements to produce audible content and bypass such interfaces? 

> For example, there could be a MediaFocusSet that you can put elements into; the default action of play() is to check if the target is in the set, and if so, do the behavior described here (pause the currently-focused element, play the new one).

The way audible/visual media is described on the web is currently through `HTMLVideoElement`, `HTMLAudioElement` (temporarily excusing Web Audio API for the reasons outlined in the [README](https://github.com/richtr/html-media-focus#what-about-the-web-audio-api)). That is to say, these interfaces provide the _essential tooling_ we need to represent media in media players - things like play, pause, stop and seeking functionality.

Using something like `MediaFocusSet` would mean we create _indirect_ code paths. In that case events originating from media control interfaces would need to wired toward any media objects attached to such a container interface. That 'wiring' would need to be performed either by the UA implicitly or left to web developers to pause and play media when requested. We could let web pages take responsibility for playing or pausing their in-page media when asked and that would work well..until one page ignores such events and the model is broken and the user blames their browser.

The other key concept is that this proposal is effectively about creating a strong _two-way binding_ (called `reflection` in the README) between media control interfaces and web media. That is to say, any changes that occur to media playback in a web page would be automatically applied to all active media control interfaces. Equally, any actions performed in media control interfaces would be automatically applied to in-page media elements. That reduces the need for web developers to take responsibility for wiring up those actions correctly and ensures we can enforce logical behaviour in the platform around media control.

> what happens if the play() event is canceled? Does the element still get media focus?

An element can only obtain focus after `playing` is fired indicating that the media has passed e.g. `canPlay` checks and the media has been successfully fetched over the network. i.e. there's a implicit promise that media playback is definitively in progress before media focus can ever be established.

> To fully explain it, you'd need to solve the Web Audio problem

This is a problem IMO but, as we stated in the [README](https://github.com/richtr/html-media-focus#what-about-the-web-audio-api), Web Audio lacks the equivalent tooling that `HTMLMediaElement`-based interfaces expose. Fixing that within Web Audio interfaces is not trivial. Leaving media control up to web developers when holding media focus could lead to buggy or plain broken user experiences.

I'd much rather we figure out a way to pass Web Audio generated content to a HTMLMediaElement-based object and then use that object to request media focus as per this proposal.

---

To re-cap: the current scope of HTML Media Focus has been chosen for its ability to create implicit two-way bindings between in-page media content and any media display interfaces. Also indireciton is avoided at all costs due to the complexity, either on a UA or a web developer in enforcing media control events get propagated and applied correctly when other media take over the media focus.

**Comment by richtr** [comment]
There were lots of typos in my previous post. I've fixed them up now.

**Comment by tabatkins** [comment]
The two-way binding is very implicit right now, though.  As described, it's just that the media keys get translated directly into media events on the focused media element (rather than the author having to handle keypresses directly), and that the media element exposes a handful of attributes that the media player can observe for use in its display.  The only explicit link is that the set of "focusable media objects" is currently defined as "the set of `<audio>` and `<video>` elements with a `focusable` attribute", which is something that we can easily open up via a number of mechanisms.

Both of the implicit connections seem completely handleable by an author with a custom object.  It's a little more work than you have to do when you can supply an `<audio>`  directly (because the event handling and attribute setting has already been coded for you), but it's not a ton of work, and it's acceptable and understandable that you have to do some extra work when you're reinventing things on your own.

**Comment by domenic** [comment]
CustomHTMLAudioElement would be implemented in terms of web audio, and CustomHTMLVideoElement in terms of canvas.

**Comment by richtr** [comment]
Taking a step back here: a more generic API would be great but if we look at platform level APIs available then implementing more a generic, permissible API is going to prove difficult across all current platforms.

If we look at [iOS Remote Control Events](https://developer.apple.com/library/ios/documentation/EventHandling/Conceptual/EventHandlingiPhoneOS/Remote-ControlEvents/Remote-ControlEvents.html#//apple_ref/doc/uid/TP40009541-CH7-SW3), for example, it is clear we cannot obtain media focus until _observable_ audio output has started within our app. Android does not share this model and is indeed more permissive in how you can grab media controls.

Any generic proposal would need to fit these platform-level restrictions. A generic proposal that lets you grab media focus before media playback has started in a web page and is not closely tied to _observable_ media playback is likely to cause problems cross-platform. Perhaps that explains why the two-way binding in this proposal is implicitly bound to ongoing, observable media playback.

What exactly _observable media_ means is debatable. I'd suggest firing image data toward a canvas at 60Hz would not constitute an observable video stream (with the kind of bindings we would need in media control interfaces).

**Comment by domenic** [comment]
> What exactly observable media means is debatable.

Is it? I thought it'd be documented in each platform's restrictions. What are they for iOS, concretely?

**Comment by richtr** [comment]
iOS requires audio to be playing out from an app for that app to acquire
the remote control events.

**Comment by tabatkins** [comment]
That doesn't seem hard to verify.  When you register for the media focus set, you can provide the WebAudioNode that'll be used for output, so the browser can verify that it's producing something.  Similarly for video - provide the canvas.

**Comment by tabatkins** [comment]
Note: I'm not (and I'm pretty sure Domenic's not) suggesting that this sort of API be what people _usually_ use.  The API sketched out in the draft currently, where you just put an attribute on some elements and everything works, is great.  It should just be explained as sugar over top of the real API, which has sufficient hooks for a custom element to use the functionality.

**Comment by domenic** [comment]
Yep. It'd be a huge shame if you could only get media controls for `<audio>` tags, but then someone makes a DJ remixing app or a programmatic-sound-generator app and the audio controls stop working.

**Comment by domenic** [comment]
Oh, and on the video side of things, a good example is x-gif.

**Comment by richtr** [comment]
This has been very productive. I'm open to specific ideas on how we could provide an additional 'media focus set'-like API. There are a lot of considerations for media controls but we should, where possible, always be providing API hooks in-line with Extensible Web principles. 

Let's leave this issue open pending solid proposals on how we could provide something more generic too.

**Comment by marcoscaceres** [comment]
Well, the challenge remains to do exactly what @domenic said:

> CustomHTMLAudioElement would be implemented in terms of web audio, and CustomHTMLVideoElement in terms of canvas.

We obviously need to start there - not at the HTML attribute (we will land at the attribute, I'm sure). 

There may be additional clues for solving this in http://w3c.github.io/media-source/  - Though, I don't know .... I'm investigating.

 We might also want to talk to @padenot about the best way to wire this up with Web Audio.  

Please be aware that time is of the essence here - Gecko is implementing the other proposal:  https://bugzilla.mozilla.org/show_bug.cgi?id=1112032

**Comment by padenot** [comment]
An AudioContext has exactly one audio output (represented by the AudioDestinationNode). There can be multiple AudioContext per page, although there is usually only one. An AudioContext can be suspended, resumed (and closed, but this is not very important here), using methods accessible from content: this would be the equivalent of .pause() and .play() of the `HTMLMediaElement`.

AudioContext is, right now, not capable of reimplementing `<audio>` in a not terribly sub-optimal way (it is somewhat planned that it would get new APIs to do so, as they would be useful in practice -- in short, one of the major issues is that Web Audio API is not capable of decoding just a chunk of compressed audio data, although there are other issues). However, @jernoble proposes some code [here](https://github.com/jernoble/Sound) that implements `HTMLMediaElement`, using today's `AudioContext` (therefore in a very sub-optimal way, but still).

In an ideal world, the stack would be layered like so (<- means "can be reimplemented in terms of):

```
??? (see last paragraph) <- AudioContext + chunk decoding capabilities or hacks <- MSE + `<audio>` <- regular `<audio src=example.org/mp3.mp3>`
```

Additionaly, the audio-wg has plans to add something like a "bag of properties" authors would pass when creating an AudioContext, much like gUM constraints, so we could use this to make it focusable:

``` js
var ac = new AudioContext({sampleRate: 44100 /* Hz */, focusable: true});
```

If we want AudioContexts to stop being focusable, we can also reflect this `focusable` property as an attribute on the AudioContext ?

In any case, cc @cwilso, maybe he'll have additional remarks/comments, maybe about the audio output stream access we keep talking about? I did not mention it here because it looks like it's not coming soon enough, but it's probably a good thing to at least consider something that would be lower level than the Web Audio API (which is a fairly high level API), as this "audio output stream object" would serve as the foundation for audio output on the web.

**Comment by cwilso** [comment]
I don't think you want AudioContexts to be focusable - because AudioContexts don't have a concept of many of the semantic messages you're talking about (play/pause/stop, etc).  I think it's important to have an underlying programmatic API - at the level of the Gecko proposal Marcos pointed to - to build on top of, so that an eventual "<audio> built on top of Web Audio" could support <audio focusable> (which is a fine semantic-level feature, but you need that bedrock feature of "can support media control events".

**Comment by jernoble** [comment]
@cwilso 

> AudioContexts don't have a concept of many of the semantic messages you're talking about (play/pause/stop, etc)

Sure they do. resume(), suspend(), and close(), in that order. But that's irrelevant. 

If an AudioContext can receive media control events, then a CustomHTMLAudioElement can be built atop it which also receives media control events.

The alternative would be to allow a web page without any audio or video capabilities to steal focus away from other pages and apps which are currently playing audio. And that's worth avoiding.

**Comment by tabatkins** [comment]
> The alternative would be to allow a web page without any audio or video capabilities to steal focus away from other pages and apps which are currently playing audio. And that's worth avoiding.

Not necessarily.  You can wait until the page is actually playing audio before giving it real media focus.  This is a detectable thing.

**Comment by jernoble** [comment]
@tabatkins 

> Not necessarily. You can wait until the page is actually playing audio before giving it real media focus. This is a detectable thing.

Which means this free-floating API will only "work" after an audio-producing object is used elsewhere on the page in a possibly entirely unrelated context.  That's awful.  Just add the API to AudioContext.

**Comment by cwilso** [comment]
@jernoble:

> > AudioContexts don't have a concept of many of the semantic messages you're talking about (play/pause/stop, etc)
> > Sure they do. resume(), suspend(), and close(), in that order. But that's irrelevant.

Not at all.  "Play" does not mean the same as resume(); "stop" doesn't mean the same thing as close().  (For example, once close() has been called, the context is effectively dead; but "stop" just means stop and reset current position to the beginning.  The AudioContext does not maintain a replayable timeline.

But regardless, yeah, this API needs to be added to AudioContext, or to window/navigator.

**Comment by jernoble** [comment]
@cwilso 

> Not at all. "Play" does not mean the same as resume();

No, not exactly, but close enough.

> "stop" doesn't mean the same thing as close(). (For example, once close() has been called, the context is effectively dead; but "stop" just means stop and reset current position to the beginning. The AudioContext does not maintain a replayable timeline.

Neither does the &lt;audio&gt; element have the "concept" of "next track", but we're talking about that semantic message as well.  In the absence of any intervention by the page, at the very least "play" and "pause" could reasonably map onto `resume()` and `suspend()`.

**Comment by jernoble** [comment]
And thinking upon this further, AudioContext has the same semantics as an audio element with an infinite duration and an empty seekableRange (i.e. a live stream).

**Comment by richtr** [comment]
RE: AudioContext and Media Control semantics: 

I also disagree that the semantics of AudioContext are synonymous with HTMLMediaElement.

> AudioContext has the same semantics as an audio element with an infinite duration and an empty seekableRange (i.e. a live stream).

Agreed.

If I could assign Web Audio to a HTMLMediaElement then we could easily adopt media control semantics such as playing and pausing over what is effectively a live streaming media source. That would mean if I pause such an HTMLMediaElement object then the audio would not longer be routed out to the speakers while the underlying AudioContext would not be harmed. It does not mean the underlying AudioContext would be suspended. It gives us a way to clearly delineate any audio processing that may be happening in AudioContext from actual output to the speakers (and hence remote controls).

If we could optionally delegate an HTMLMediaElement object to be the audio output sink for an AudioContext object then we could get to the point where Web Audio processing and Web Audio playback could be decoupled in-page while defining a common audio output sink that happens to implement the required semantics we need to HTML Media Focus.

**Comment by richtr** [comment]
(re-written for clarity...)

RE: AudioContext and Media Control semantics: 

I also disagree that the playback semantics of AudioContext are synonymous with HTMLMediaElement. i.e. I think `play() !== resume()` and `pause() !== suspend()`. If it did then we would have wrapped AudioContext with the same attribute-based, platform-enforceable proposal.

> AudioContext has the same semantics as an audio element with an infinite duration and an empty seekableRange (i.e. a live stream).

If we could delegate an HTMLMediaElement object to be the audio output sink for an AudioContext 'live stream' object then we could get to the point where Web Audio processing and Web Audio playback could be decoupled in-page. We would also be enabling a common output sink for audio on the web that happens to implement the required semantics we need for remote control events.

If I paused such an AudioContext-bound HTMLMediaElement object then the audio would simply no longer be routed out to the speakers but the underlying AudioContext would not be harmed. It would not mean the underlying AudioContext should be suspended (though the element may suspend its AudioContext object when 'paused' or 'ended' events are received on the audio output element). This gives us a route to delineate **audio processing** that may be happening in AudioContext from actual **audio output** to the speakers and solves the issue of allowing AudioContext to obtain remote control focus while also making sure it implements the semantics we need in remote control interfaces.

This is per the proposal we made in the [README](https://github.com/richtr/html-media-focus/blob/gh-pages/README.md#what-about-the-web-audio-api).

**Comment by jernoble** [comment]
@richtr 

> If we could delegate an HTMLMediaElement object to be the audio output sink for an AudioContext 'live stream' object then we could get to the point where Web Audio processing and Web Audio playback could be decoupled in-page. 

This is an interesting proposal. WebAudio already has the concept of using HTMLMediaElement as a source, with [`MediaElementAudioSourceNode`](http://webaudio.github.io/web-audio-api/#MediaElementAudioSourceNode), and of a non-device sink, with [`MediaStreamAudioDestinationNode`](http://webaudio.github.io/web-audio-api/#the-mediastreamaudiodestinationnode-interface). So under your original proposal, it should already be possible to make a "focusable" AudioContext by [creating a `HTMLAudioElement`](https://dvcs.w3.org/hg/audio/raw-file/tip/streams/StreamProcessing.html#audio-element-extensions) with `MediaStreamAudioDestinationNode`.

Of course, you couldn't mandate that HTMLAudioElements were the only way to produce audio on the web platform without making extensive, breaking changes to the Web Audio API, but it could be an opt-in mechanism for "focusable" AudioContexts. But:

> This gives us a route to delineate audio processing that may be happening in AudioContext from actual audio output to the speakers and solves the issue of allowing AudioContext to obtain remote control focus while also making sure it implements the semantics we need in remote control interfaces.

I don't necessarily agree that this would magically make all the semantics match. Under this proposal, a "pause" command effectively turns into "mute", for example. And "seek", "rewind", "fast forward", "next/previous track", "stop" commands would still fail. You could get the exact same semantics by speccing that "focusable" AudioContexts must mute themselves when they receive the "pause" command.

**Comment by marcoscaceres** [comment]
@domenic, elsewhere I proposed the following:

> Just thinking out loud, but something like where both `RemoteControl` and `HTMLMediaElement` implement [NoInterfaceObject] `MediaKeyReceiver`. 
> 
> This would allow MediaKeyReceiver to remain as an independent API, but HTMLMediaElement could still be explained (as far as media-keys are concerned) in terms of `MediaKeyReceiver`. A constructable, or requested, `RemoteControl` will hopefully then allow the ability to address use cases like controlling RevealJS presentations, the Flash use cases, and things built with custom elements, etc. 
> 
> `RemoteControl` would have to be requested (`requestRemoteControl()`), granting the user the explicit ability to now allow arbitrary web apps to claim the media focus if they are not actually controlling media (or if they are playing media through a third-party plugin, as per Flash). Hopefully requiring an explicit grant should also prevent media-key hijacking.   

Would something like that work?

**Comment by richtr** [comment]
@marcoscaceres In an effort to give detailed feedback on [your proposal above](https://github.com/richtr/html-media-focus/issues/1#issuecomment-73689451) I have created issue [#5 - `Proposal to separate media focus from HTMLMediaElement objects`](https://github.com/richtr/html-media-focus/issues/5). Is that OK?
